devlib/instrument/baylibre_acme.py

```diff
@@ -0,0 +1,424 @@
+            raise OSError('Connection to the ACME timed out. This object is '
+                          'now invalid and should be destroyed. The ACME may '
+                          'also require a reboot.')
+
+        self.data = collections.OrderedDict()
+        for ch in self.active_channels:
+            self.data[ch.label] = ch.iio_get_samples(self.absolute_timestamps)
+            ch.iio_forget_samples()
+
+    def get_data(self, outfile=None):
+        if not outfile:
+            return self.data
+
+        with open(outfile, 'wb') as f:
+            w = csv.writer(f)
+            w.writerow(self.data.keys())
+            w.writerows(zip(*self.data.itervalues()))
+        return MeasurementsCsv(outfile, self.active_channels, self.sample_rate_hz)
+
+    def get_raw(self):
+        # TODO How should this be implemented knowing that we don't output each
+        #      channel to its own CSV?
+        return []
+
+
+def _BaylibreAcmePropertyFromProbes(property_name, none_accepted=False):
+
+    def getter(self):
+        active = hasattr(self, 'active_probes') and self.active_probes
+        probes = self.active_probes if active else self.probes
+        try:
+            # This will fail if there is not exactly one single value:
+            (value,) = {getattr(probes[p], property_name) for p in probes}
+        except ValueError:
+            msg = 'Probes have different values for {}.'
+            raise ValueError(msg.format(property_name))
+        return value
+
+    def setter(self, value):
+        # This setter is required for compliance with the inherited methods
+        if value is not None:
+            raise AttributeError("can't set attribute")
+
+    return property(getter, setter if none_accepted else None)
+
+class BaylibreAcmeInstrument(Instrument):
+
+    def __init__(self, iio_context=None, use_base_iio_context=False, probe_names=None):
+        if isinstance(probe_names, basestring):
+            probe_names = [probe_names]
+
+        self.iio_context = (iio_context if not use_base_iio_context
+                            else iio.Context(iio_context))
+
+        if probe_names is not None:
+            if len(probe_names) != len(self.iio_context.devices):
+                msg = ('There should be as many probe_names ({}) '
+                       'as detected probes ({}).')
+                raise ValueError(msg.format(len(probe_names),
+                                            len(self.iio_context.devices)))
+
+            probes = [IIOINA226Instrument(d, name)
+                      for d, name in zip(self.iio_context.devices, probe_names)]
+        else:
+            probes = [IIOINA226Instrument(d) for d in self.iio_context.devices]
+
+        self.probes = {p.nickname : p for p in probes}
+
+    # channel management
+
+    def list_channels(self):
+        # TODO: not sure what to do:
+        pass
+        #      or
+        # return {p: self.probes[p].list_channels() for p in self.probes}
```

Comment:
If we were to implement the output as a `dict`, it wouldn't comply with `Instrument`.

If we decided to output channels such as `'<probe>/<site>_<kind>'`, we can't prepend `p` to the channel name in this function (I believe) so that the channel itself should be able to output that string. A way would be to overwrite its `'<site >'` with `'<probe>/<site>'` but then `IIOINA226Instrument.reset` could not rely on the corresponding method of its parent ... Am I missing something?

This is an obvious drawback of my design choice of using an `Instrument` to hold `Instrument`s ...

Reply:
I would question the need for the `IIOINA226Instrument` to be an actual `Instrument`. Each probe is one <site> exposing timestamp, voltage (I don't see a reason to expose both bus and shunt voltages to the outside world), current, and power. `IIOINA226Instrument` is basically an interface to a probe, but it does not need to implement the `Instrument` API, as that just makes things more complicated and confusing.